docs/spec/special-types.rst

```diff
@@ -113,11 +113,25 @@ Special cases for ``float`` and ``complex``
 
 Python's numeric types ``complex``, ``float`` and ``int`` are not
 subtypes of each other, but to support common use cases, the type
-system contains a straightforward shortcut:
-when an argument is annotated as having
-type ``float``, an argument of type ``int`` is acceptable; similar,
-for an argument annotated as having type ``complex``, arguments of
-type ``float`` or ``int`` are acceptable.
+system contains a special case.
+
+When a reference to the built-in type ``float`` appears in a :term:`type expression`,
+it is interpreted as if it were a union of the built-in types ``float`` and ``int``.
+Similarly, when a reference to the type ``complex`` appears, it is interpreted as
+a union of the built-in types ``complex``, ``float`` and ``int``.
+These implicit unions behave exactly like the corresponding explicit union types,
+but type checkers may choose to display them differently in user-visible output
+for clarity.
+
+Type checkers should support narrowing the type of a variable to exactly ``float``
+or ``int``, without the implicit union, through a call to ``isinstance()``::
```

Comment:
IIUC this is basically saying that if `float` appears as an argument to `isinstance()` then we should interpret it differently from the case when `float` appears in function annotations?

In that case, I'm wondering about what should happen in other cases, e.g. `cast()`, `assert_type()`, or other kinds of type forms -- should we interpret `float` to be the "real" float in those places as well? Does it make sense to have the spec discuss about the "scope" of this kind of special-casing interpretations?

Reply:
The proposed wording says that `float` should be interpreted as `float | int` in type expressions. The first argument to `cast()` is a type expression, but the second argument to `isinstance()` is not.